Add upgraded block header type

ss-es opened the pull request.

sveitser commented:
I don't know if this would have some knock-on effects but instead of changing the type of a field should we make a new field like `timestamp_ns` instead, or add a separate fields for the nanoseconds to maintain backwards compatibility for the original timestamp? I feel like that might be a bit less confusing than changing the meaning of timestamp values by changing their unit. I'm afraid this will be a subtle breakage for all consumers that may go unnoticed for a while and potentially cause downtime.

Also do we need it to be nanoseconds and u128? Unless I made a mistake we still have only used a 1/10 of the u64 nanoseconds since 1970. In system with about 2 seconds latency it seems to me that more than 1s resolution isn't very meaningful and we don't have a way to enforce the value to be very accurate, even now with 1 second resolution. I forgot why exactly we want more resolution do we really need to do this in the header?

cc @ss-es @bfish713

sveitser commented:
We should also add reference tests in `data/v4`.

bfish713 commented:
> I don't know if this would have some knock-on effects but instead of changing the type of a field should we make a new field like `timestamp_ns` instead, or add a separate fields for the nanoseconds to maintain backwards compatibility for the original timestamp? I feel like that might be a bit less confusing than changing the meaning of timestamp values by changing their unit. I'm afraid this will be a subtle breakage for some consumers that may go unnoticed for a while and potentially cause downtime.
> 
> Also do we need it to be nanoseconds and u128? Unless I made a mistake we still have only used a 1/10 of the u64 nanoseconds since 1970. In system with about 2 seconds latency it seems to me that more than 1s resolution isn't very meaningful and we don't have a way to enforce the value to be very accurate, even now with 1 second resolution. I forgot why exactly we want more resolution and just wondering if we really need to do it via the header.
> 
> cc @ss-es @bfish713

The reason we want more resolution is for benchmarking.  We often see rounds taking < 1s, especially in devnets.  It's very possible we introduce a change that reduces total latency by 300ms and we would have a hard time seeing it with 1s resolution.  But the main reason is that we want to get latency down to sub 1s anyway, at that point the 1s resolution will really not be enough. We want to have this via the header because we want to at least estimate the total time it takes a proposal from creation to decide. There could be some timestamp drift but that should be quite low for most running in the cloud.

Completely agree with the first part, we should add the new timestamp rather than replace the old.

ss-es commented:
thank you very much for the suggestion @sveitser, I've updated the PR to add a new field and I think this solved a lot of the trouble I was having! I was also very uncomfortable with the way I was approaching it previously -- it was too easy to make undetectable mistakes

I've also added reference tests for the new header type

For the type, I was using `u128` initially to be closer to the type from the crate: https://docs.rs/time/latest/time/struct.OffsetDateTime.html#method.unix_timestamp_nanos. I was mostly trying to avoid having to do a conversion here and I've updated the type to `i128` to avoid the conversion entirely. but if there are strong opinions on this, I'm open to changing it

imabdulbasit reviewed (commented):
Could you please include the genesis.toml file for the new v4 version, as well as the upgrade from v3 to v4, so that we can run the native demo locally?

ss-es commented on types/src/v0/impls/state.rs:
```diff
@@ -407,6 +412,18 @@ impl<'a> Proposal<'a> {
 
+    /// The `timestamp` and `timestamp_nanos` fields must be coherent
+    fn validate_timestamp_consistency(&self) -> Result<(), ProposalValidationError> {
+        if self.header.timestamp() as i128 * 1_000_000_000 != self.header.timestamp_nanos() {
```
oops -- this should be `timestamp_nanos / 10^9` on the other side

imabdulbasit commented on sequencer/src/run.rs:
```diff
@@ -29,6 +29,35 @@ pub async fn main() -> anyhow::Result<()> {
     let upgrade = genesis.upgrade_version;
 
     match (base, upgrade) {
+        #[cfg(all(feature = "fee", feature = "pos"))]
```
There should be a new feature for this upgrade?

imabdulbasit replied:
and also sequencer-sqlite
